### DISCUSSION TITLE
Nuxt 3 Authentication with useCookie

### DISCUSSION THREAD

**Discussion by lacorde** [nuxt/nuxt#16429] [Questions]
According to [this tweet](https://twitter.com/Atinux/status/1557803695080980480?s=20&t=2ND_t4gcaW87RcKKVCXFXQ) from @Atinux, Nuxt3 authentication can easily be achieved with `useCookie()` composable.

I'm quite new to Nuxt 3 (and authentication in general), and I don't know where to start.
Is there a comprehensive guide/post/stackblitz somewhere that could help me (and others) implement this **with a JWT token** ?

Thank you :)

Edit: nuxt/framework#6804 answers the question partially, some more indications could be of use !

**Comment by fathi91** [comment]
Hi,
i have implemented a JWT token authentication and it is very easy to do in nuxt 3 you just need to understand the steps
i'll share a working code
the first thing you need is to register a new user and hash the password, i use for that `argon2` you can use whatever you want
second thing you need to be able to generate and validate the **JWT tokens** and i use `jsonwebtoken` for that

here is a snippet to register a new user
in `server/api/user.ts` 
**Note that i use prisma as my ORM**
```javascript
import { dbClient } from "@/server/utils/db-client";
import argon2 from "argon2";
export default defineEventHandler(async (e) => {
  let method = getMethod(e);
if (method == "POST") {
    let body = await readBody(e);
    let { password } = body;
    let hash = await argon2.hash(password);
    try {
      await dbClient.user.create({
        data: {
          name: body.name,
          email: body.email,
          password: hash,
          role: body.role,
        },
      });
      e.res.statusCode = 201;
      e.res.end();
    } catch (err) {
      console.log(err);
      e.res.statusCode = 500;
      e.res.end();
    }
  }
});
```
so after you register a new user you need now to implement a login route
in `server/api/login.post.ts`

first check if the user credentials correct and then generate the JWT token and use `setCookie` to set the cookies with the response
```javascript
import { dbClient } from "@/server/utils/db-client";
import getToken from "@/server/utils/generate-signed-token";
import argon2 from "argon2";

export default defineEventHandler(async (e) => {
  const body = await readBody(e);
  let { email, password } = body;
  let user: any;
  try {
    user = await dbClient.user.findFirst({
      where: {
        email: email,
      },
      include: {
        role: true,
      },
    });
  } catch (e) {
    console.log(e);
    throw createError({ statusCode: 500, statusMessage: "internal server error" });
  }
  let resData = null;
  let resMsg = "Email or password incorrect.";
  let resStatus = false;
  if (user && (await argon2.verify(user.password, password))) {
    let now = new Date();
    now.setHours(now.getHours() + 1);
    let authtoken = await getToken(user);
    setCookie(e, "Authentication", authtoken["Authentication"], { expires: now), httpOnly: true, sameSite: true });
    let { record_id, name, email, role } = user;
    resData = { id: record_id, name, email, role: { name: role.name, src: Object.keys(role.table_source) } };
    resMsg = "Success";
    resStatus = true;
    return { status: resStatus, message: resMsg, data: resData };
  } else {
    return { status: resStatus, message: resMsg, data: resData };
  }
});
```
in `server/utils/generate-signed-token.ts`
```javascript
import jwt from "jsonwebtoken";

export default async function getToken(payload: any) {
  const runtimeConfig = useRuntimeConfig();
  let { record_id: id, role } = payload;

  let token = jwt.sign({ id: id, role }, runtimeConfig.JWT_KEY);

  return {
    Authentication: runtimeConfig.JWT_PREFIX + " " + token,
  };
}
```

and if you want to logout the use you just simply delete the cookies like this
```javascript
export default defineEventHandler((e) => {
    e.res.writeHead(302, { Location: '/login', 'Set-Cookie': 'Authentication=; Max-Age=0; Path=/; HttpOnly; SameSite=Strict' })
    e.res.end()
})
```

so now the server part is done here is how you could integrate this with the frontend using nuxt plugin
in `plugins/auth.ts`

```javascript
import { IUser } from "@/types";
export default defineNuxtPlugin(async (NuxtApp) => {
  let user = useState<IUser | null>("user", () => null);
  let isLoggedIn = useState<boolean>("isLoggedIn", () => false);
  if (process.server) {
    try {
      user.value = await $fetch("/api/me", {
        headers: useRequestHeaders(["cookie"]) as any,
      });
      isLoggedIn.value = true;
    } catch (e) {}
  }

  let logIn = async (email: string, password: string) => {
    let res = await $fetch("/api/login", { method: "POST", body: { email, password }, headers: useRequestHeaders(["cookie"]) as any });
    return res;
  };
  let logOut = async () => {
    let res = await $fetch.raw("/api/logout");
    user.value = null;
    isLoggedIn.value = false;
    window.location.href = res.url;
  };

  return {
    provide: {
      auth: { logIn, user, logOut, isLoggedIn },
    },
  };
});
```

then use the plugin in your components
```javascript
<script setup>
const { $auth } = useNuxtApp();
let { logIn, isLoggedIn, user } = $auth;

/* now you can use isLoggedIn to cehck if the user already logged-in or you can login or logout the user using logIn, logout functions
<script/>
```
just like that you should have a basic authentication system, only one thing is missing
now when the client sends a request with cookies you need to verify if the user is authenticated or not
we can use server middleware for that
in `server/middleware/auth.ts` you can verify the JWT token
```javascript
export default defineEventHandler(async (e) => {
  let cookies = parseCookies(e);
 // first check if there is cookies with the request if not throw  an error the user not authenticated
  if (!cookies.Authentication) {
    throw createError({ statusCode: 401, statusMessage: "Unauthorized" });
  }
 
  //decode the JWT token with the same key that was created with
  // I have the key in the runtime configration
  let runtimeConfig = useRuntimeConfig();
  let authData = cookies.Authentication.split(" ");
  let tokenPrefix = authData[0];
  if (tokenPrefix !== runtimeConfig.JWT_PREFIX) {
    throw createError({ statusCode: 401, statusMessage: "Unauthorized" });
  }
  let token = authData[1];

  let decodedPayload: any;
  try {
    decodedPayload = jwt.verify(token, runtimeConfig.JWT_KEY);
  } catch (e) {
    console.log(e);
    throw createError({ statusCode: 503, statusMessage: "Service Unavailable" });
  }
 //after you successfully decode and  verify the JWT token you can add the user in the request context to use it in the requested route
try {
      let user = await dbClient.user.findFirst({
        where: {
          record_id: decodedPayload.id,
        },
        select: {
          name: true,
          email: true,
          record_id: true,
          is_active: true,
          role: true,
        },
      });
      e.context.user = user;
    } catch (e) {
      console.log(e);
      throw createError({ statusCode: 503, statusMessage: "Service Unavailable" });
    }
```

**obviously you need a better error handling and i guess the code could be better, this was just an example**
i hope that helps

**Reply by lacorde** [reply]
Thank you so much for your detailed response !
I will look into this and tell you if it fits my needs :)

**Comment by BracketJohn** [comment]
Hey @lacorde! 

We had the same problem of authentication & user sessions, so we published two modules for it:
- [nuxt-auth](https://github.com/sidebase/nuxt-auth): End-to-end authenticstion library based on the popular NextAuthjs library. Use it to get almost zero config authentication
- [nuxt-session](https://github.com/sidebase/nuxt-session): Vanilla sessions that persist user data across requests. Use it to roll your own auth if you want to

Generally, Id recommend not to implement auth from the ground up, as it has many intricacies that one can get wrong (: Hope that one of these can help you to implement authentication without having to code too much of it by yourself 🙌

**Comment by RoyalVICTOR-pro** [comment]
Hi @fathi91 ! Hi @BracketJohn !

I have a slightly different scenario, and I would like to know if your solutions would be suitable. 

I've developed an API using AdonisJS (NodeJS) as my backend. 
I'm working on a Single Page Application with Nuxt 3 (Universal mode with Composition API) that requires users to be logged in. 
When a user logs in, the API returns an OAT (Opaque Access Token). 
I want to store this OAT in a Secure and HttpOnly cookie. 
This OAT should be sent with every request to the API to verify that the user is indeed logged in and has access to specific resources.

As I'll have mobile applications connecting to my API, and having learned that storing tokens in localStorage or sessionStorage is not secure, that's why I want to use OATs stored in secure and HttpOnly cookies.

What would you advise me? Do you have any resources to recommend (article, video, course...)?

Thanks !!

**Comment by RoyalVICTOR-pro** [comment]
For now, I've done something like this:

![2023-12-29 13_10_36-login vue - steewo-backend-app - Visual Studio Code](https://github.com/nuxt/nuxt/assets/145127368/9e3d1e9d-5b19-4ba3-bed8-d0c3989c878c)

But when, after logging in, in my authStore.ts at line 22, I create a cookie with the retrieved token and redirect to the requested page (in login.vue at line 32), the middleware that checks if I have access to the requested page does not have the cookie. (When I log on the API side, I see that the cookie with the token is not present.)
